### ISSUE TITLE
Wrong linenumber reported for %typemap unknown baseclass warning

### ISSUE THREAD

**Issue by ojwb** [swig/swig#2781]
```
$ cat wronglinenumber.i 
%module x
namespace A {
  class XYZ {};
  template<typename T> struct ABC : public B::ABC<T> {};
}
namespace B {
  template<typename T> struct ABC {
    void aaa(T t) {}
  };
}
%template(ABCXYZ) A::ABC<A::XYZ>;
$ ../preinst-swig -python -c++ wronglinenumber.i 
wronglinenumber.i:1: Warning 401: Nothing known about base class 'B::ABC< A::XYZ >'. Ignored.
wronglinenumber.i:1: Warning 401: Maybe you forgot to instantiate 'B::ABC< A::XYZ >' using %template.
```

The line number of these warnings seems completely wrong - I would probably expect the warnings to be reported at line 4 or perhaps line 11.

Looking at the code, these warnings come from a single place which reports the location as `Getfile(bname), Getline(bname)` where `bname` is the base class type.  I couldn't work out where this gets set.

**Comment by wsfulton** [comment]
The line numbers are set in base_specifier in the parser
https://github.com/swig/swig/blob/2823c70096863a32b4d45e26225ef3b618ca6c56/Source/CParse/parser.y#L6956-L6957

Line numbering work for non-template classes, so probably this info is being lost when instantiating the template. I'll have a look, it shouldn't be hard to find.

**Comment by ojwb** [comment]
Thanks.

I was curious about why it ended up as `1` - looking at the implementation of line number tracking in DOH, `Hash` and `List` default to `line = 0` which `String` defaults to `line = 1`:

```
Source/DOH/hash.c:  h->line = 0;
Source/DOH/list.c:  l->line = 0;
Source/DOH/string.c:  s->line = 1;
Source/DOH/string.c:  str->line = 1;
Source/DOH/string.c:  str->line = 1;
Source/DOH/string.c:  str->line = 1;
```

I think this ought to be consistent (and `0` seems more logical for "not set yet" as that's not a valid line number).

Also seems odd that `String_clear()` resets `s->line = 1;` but doesn't reset `file`, while `Hash_clear()` and `List_clear()` don't reset `line` or `file`.

Or are there reasons I'm not spotting for these differences?

**Comment by wsfulton** [comment]
I've also spotted that String_clear does not clear everything, file in particular. That is part of what I'm cleaning up. The numbering is lost with a Clear() in SwigType_typename_replace. I'll also have a look at the default line numbering being consistent, it's a good spot.

**Comment by wsfulton** [comment]
Fixed in 8e86058c5d1f. Also see 4cb2b253d0aba4b7083d66e39d3eebeabaa88e45.
